### PR TITLE
feat(sdk): add support program registration

### PR DESCRIPTION
**Pull request by mattstam**
Problem: uploading the elf and doing prover.setup(elf) to get the vk on every request is unnecessary.

Solution: the elf and vk are no longer sent every request, instead they are only sent if the program hasn't been registered in the network yet.

Previous SDK prove:
1. upload the elf
2. call RequestProof(program_uri, vk ...) , which returns request_id
3. poll GetProofRequestStatus(request_id)

New SDK prove:
1. call GetProgram(vk_hash) , which returns some metadata about an existing program (e.g. program_uri, name) or error if doesn't exist
a. if error:
i. upload the elf
ii. call CreateProgram(vk_hash, vk, program_uri)
2. call RequestProof(vk_hash, ...), which returns request_id (fails if vk_hash doesn't exist)
3. poll GetProofRequestStatus(request_id)

---

Example output with registration, since the program doesn't exist yet:
```
2024-12-05T01:45:09.676237Z  INFO Client circuit version: v3.0.0    
vk: "0x00aa4ee3c793869d5023ae21800c827d39a23a8c594f07729dd6391b7091a89d"
2024-12-05T01:45:11.107286Z  WARN network_prover: close time.busy=8.44ms time.idle=2.58µs  
2024-12-05T01:45:11.407481Z  INFO Registered program 0x00aa4ee3c793869d5023ae21800c827d39a23a8c594f07729dd6391b7091a89d    
2024-12-05T01:45:11.407494Z  INFO Requesting proof with cycle limit: 100000000    
2024-12-05T01:45:11.556967Z  INFO Created request 0x1c1dafab91a6a7a829ab3d0faea0733b8136e2a6037e91024a526833ab7a023a in transaction 0xd0d15547d3c047ea8c62cdadd229ceb1329c6fa8f5c728decbaaa0581a3a123e    
2024-12-05T01:45:11.561052Z  INFO Proof request assigned, proving... 
generated proof
public values: 0xf4010000f404000086070000
proof: 0x090...
```
and then re-running it, it skips the registration:
```
2024-12-05T02:12:12.128586Z  INFO Client circuit version: v3.0.0    
vk: "0x00aa4ee3c793869d5023ae21800c827d39a23a8c594f07729dd6391b7091a89d"
2024-12-05T02:12:13.556157Z  WARN network_prover: close time.busy=10.9ms time.idle=2.92µs
2024-12-05T02:12:13.566362Z  INFO Requesting proof with cycle limit: 100000000    
2024-12-05T02:12:13.760741Z  INFO Created request 0x83d062654a3ef75c34b2338e32fb51e39586f6f498f29b6381f8c773725f3e30 in transaction 0x99c6b3bd36124c8289e95c36df0522d94f4623c995aebf9b6b99a6049bd07064    
2024-12-05T02:12:13.764281Z  INFO Proof request assigned, proving... 
generated proof
public values: 0xf4010000f404000086070000
proof: 0x090...
```

### CODE REVIEW

**Comment by mattstam** [comment]
> error: rustc 1.79.0 is not supported by the following packages:
  alloy-primitives@0.8.14 requires rustc 1.81
  alloy-sol-macro@0.8.14 requires rustc 1.81
  alloy-sol-macro-expander@0.8.14 requires rustc 1.81
  alloy-sol-macro-input@0.8.14 requires rustc 1.81
  alloy-sol-types@0.8.14 requires rustc 1.81
  syn-solidity@0.8.14 requires rustc 1.81
Either upgrade rustc or select compatible dependency versions with
`cargo update <name>@<current-ver> --precise <compatible-ver>`
where `<compatible-ver>` is the latest version supporting rustc 1.79.0

unrelated CI failing

File: crates/sdk/src/network-v2/prover.rs [hunk 1]
```diff
@@ -38,63 +42,80 @@ impl NetworkProver {
...
         let local_prover = CpuProver::new();
         let client = NetworkClient::new(private_key, rpc_url);
-        Self { client, local_prover, skip_simulation }
+        Self { client, local_prover, skip_simulation, strategy: FulfillmentStrategy::Hosted }
+    }
+
+    /// Sets the fulfillment strategy for the client. By default, the strategy is set to `Hosted`.
+    pub fn with_strategy(&mut self, strategy: FulfillmentStrategy) {
+        self.strategy = strategy;
     }
```
Comment (mattstam) [inline crates/sdk/src/network-v2/prover.rs:51]: note: I will do other SDK improvements in a subsequent PR

File: crates/sdk/src/network-v2/client.rs [hunk 2]
```diff
@@ -95,6 +97,76 @@ impl NetworkClient {
         Ok(res.into_inner().nonce)
     }
 
+    /// Get the verifying key hash from a verifying key. The verifying key hash is used to identify
+    /// a program.
+    pub fn get_vk_hash(vk: &SP1VerifyingKey) -> Result<Vec<u8>> {
```
Comment (jtguibas) [inline crates/sdk/src/network-v2/client.rs:102]: Wondering if we should use SP1VerifyingKeyHash([u32; 8]) or something

**Review by jtguibas — APPROVED** [review]

File: crates/sdk/src/network-v2/client.rs (hunk 2 above)
Comment (mattstam) [reply crates/sdk/src/network-v2/client.rs:102]: will keep this in mind

**Pull request merged**